node/core/av-store/src/lib.rs

```diff
@@ -580,82 +617,107 @@ async fn process_block_activated(
 ) -> Result<(), Error> {
 	let now = subsystem.clock.now()?;
 
-	let candidate_events = {
+	let block_header = {
 		let (tx, rx) = oneshot::channel();
-		ctx.send_message(
-			RuntimeApiMessage::Request(activated, RuntimeApiRequest::CandidateEvents(tx)).into()
-		).await;
-
-		rx.await??
-	};
 
-	let block_number = {
-		let (tx, rx) = oneshot::channel();
 		ctx.send_message(
-			ChainApiMessage::BlockNumber(activated, tx).into()
+			ChainApiMessage::BlockHeader(activated, tx).into()
 		).await;
 
 		match rx.await?? {
 			None => return Ok(()),
 			Some(n) => n,
 		}
 	};
+	let block_number = block_header.number;
 
-	let block_header = {
-		let (tx, rx) = oneshot::channel();
+	let new_blocks = util::determine_new_blocks(
+		ctx.sender(),
+		|hash| -> Result<bool, Error> {
+			Ok(subsystem.known_blocks.is_known(hash))
+		},
+		activated,
+		&block_header,
+		subsystem.finalized_number.unwrap_or(block_number.saturating_sub(1)),
+	).await?;
 
+	let mut tx = DBTransaction::new();
+	for (hash, header) in new_blocks {
+		process_new_head(
+			ctx,
+			&subsystem.db,
+			&mut tx,
+			&subsystem.config,
+			&subsystem.pruning_config,
+			now,
+			hash,
+			header,
+		).await?;
+		subsystem.known_blocks.insert(hash, block_number);
+	}
+	subsystem.db.write(tx)?;
+
+	Ok(())
+}
+
+async fn process_new_head(
+	ctx: &mut impl SubsystemContext,
+	db: &Arc<dyn KeyValueDB>,
+	db_transaction: &mut DBTransaction,
+	config: &Config,
+	pruning_config: &PruningConfig,
+	now: Duration,
+	hash: Hash,
+	header: Header,
+) -> Result<(), Error> {
+
+	let candidate_events = {
+		let (tx, rx) = oneshot::channel();
 		ctx.send_message(
-			ChainApiMessage::BlockHeader(activated, tx).into()
+			RuntimeApiMessage::Request(hash, RuntimeApiRequest::CandidateEvents(tx)).into()
 		).await;
 
-		match rx.await?? {
-			None => return Ok(()),
-			Some(n) => n,
-		}
+		rx.await??
 	};
 
-	// We need to request the number of validators based on the parent state, as that is the number of validators
-	// used to create this block.
+	// We need to request the number of validators based on the parent state,
+	// as that is the number of validators used to create this block.
 	let n_validators = {
 		let (tx, rx) = oneshot::channel();
 		ctx.send_message(
-			RuntimeApiMessage::Request(block_header.parent_hash, RuntimeApiRequest::Validators(tx)).into()
+			RuntimeApiMessage::Request(header.parent_hash, RuntimeApiRequest::Validators(tx)).into()
```

Comment:
this looks very inefficient
we could query a session index and session info instead

Reply:
`RollingSessionWindow` would work for this, I think.

Reply:
was thinking the same and/or doing #2987